symbolic-cabi/src/sourcemapcache.rs

```diff
@@ -0,0 +1,197 @@
+use std::ops::Deref;
+use std::os::raw::c_char;
+use std::ptr;
+use std::slice;
+
+use crate::utils::ForeignObject;
+use crate::SymbolicStr;
+
+use symbolic::common::AsSelf;
+use symbolic::common::ByteView;
+use symbolic::common::SelfCell;
+use symbolic::sourcemapcache::SourceLocation;
+use symbolic::sourcemapcache::{
+    ScopeLookupResult, SourceMapCache, SourceMapCacheWriter, SourcePosition,
+};
+
+struct Inner<'a> {
+    cache: SourceMapCache<'a>,
+}
+
+impl<'slf, 'a: 'slf> AsSelf<'slf> for Inner<'a> {
+    type Ref = Inner<'slf>;
+
+    fn as_self(&'slf self) -> &Self::Ref {
+        self
+    }
+}
+
+pub struct OwnedSourceMapCache<'a> {
+    inner: SelfCell<ByteView<'a>, Inner<'a>>,
+}
+
+/// Represents an sourcemapcache.
+pub struct SymbolicSourceMapCache;
+
+impl ForeignObject for SymbolicSourceMapCache {
+    type RustObject = OwnedSourceMapCache<'static>;
+}
+
+#[repr(C)]
+pub struct SymbolicStrVec {
+    pub strs: *mut SymbolicStr,
+    pub len: usize,
+}
+
+impl SymbolicStrVec {
+    pub fn from_vec(vec: Vec<&str>) -> Self {
+        let mut strs: Vec<_> = vec.into_iter().map(SymbolicStr::new).collect();
+        strs.shrink_to_fit();
+        let rv = SymbolicStrVec {
+            strs: strs.as_mut_ptr(),
+            len: strs.len(),
+        };
+        std::mem::forget(strs);
+        rv
+    }
+}
+
+/// Represents a single token after lookup.
+#[repr(C)]
+pub struct SymbolicSmTokenMatch {
+    /// The line number in the original source file.
+    pub line: u32,
+    /// The column number in the original source file.
+    pub col: u32,
+    /// The path to the original source.
+    pub src: SymbolicStr,
+    /// The name of the function containing the token.
+    pub function_name: SymbolicStr,
+
+    pub pre_context: SymbolicStrVec,
+    pub context_line: SymbolicStr,
+    pub post_context: SymbolicStrVec,
+}
+
+ffi_fn! {
+    /// Creates an sourcemapcache from a given minified source and sourcemap contents.
+    ///
+    /// This shares the underlying memory and does not copy it if that is
+    /// possible.  Will ignore utf-8 decoding errors.
+    unsafe fn symbolic_sourcemapcache_from_bytes(
+        source_content: *const c_char,
+        source_len: usize,
+        sourcemap_content: *const c_char,
+        sourcemap_len: usize,
+    ) -> Result<*mut SymbolicSourceMapCache> {
+        let source_slice = slice::from_raw_parts(source_content as *const _, source_len);
+        let sourcemap_slice = slice::from_raw_parts(sourcemap_content as *const _, sourcemap_len);
+
+        let writer = SourceMapCacheWriter::new(
+            String::from_utf8_lossy(source_slice).deref(),
+            String::from_utf8_lossy(sourcemap_slice).deref()
+        )?;
+        let mut buffer = Vec::new();
+        writer.serialize(&mut buffer)?;
+
+        let byteview = ByteView::from_vec(buffer);
```

Comment:
This holds the sourcemapcache buffer in memory. Not sure its worth putting in the API to write it to disk, and to read/map it from disk right now. But we should think about that rather sooner than later.